basic.go

```diff
@@ -104,22 +104,38 @@ func (s *sub) Out() <-chan interface{} {
 }
 
 func (s *sub) Close() error {
-	close(s.ch)
+	stop := make(chan struct{})
+	go func() {
+		for {
+			select {
+			case <-s.ch:
```

Comment:
How about we block Close until the channel is closed and empty? We can drain the channel concurrently in a goroutine (like here) while we remove it from nodes, but we block the return until the channel is empty and closed. Makes for more predictable behaviour. We should also add a subscriber option so the consumer can decide if they want pending items drained or not on closure (this would enable certain test scenarios with for range ch).

Reply:
Removing stuff from nodes happens under write lock, which we can only take if the channel is drained.

We don't really care to drain the channel fully (if buffered), we only need to make sure we remove it from relevant nodes, and let GC take care of the rest

Reply:
IMO let's merge it, it is better than it was. If there is an edge case we will discover it.

Reply:
I'd also rather fully drain the channel, mostly because it entirely removes the "stop" channel, but either way works. let's get this fixed.